drivers/flash/soc_flash_mcux.c

```diff
@@ -89,14 +110,116 @@ static int flash_mcux_write(struct device *dev, off_t offset,
 	addr = offset + priv->pflash_block_base;
 
 	key = irq_lock();
-	rc = FLASH_Program(&priv->config, addr, (uint8_t *) data, len);
+	rc = FLASH_Program(&priv->config, (u32_t)addr, (u8_t *)data, len);
 	irq_unlock(key);
 
 	k_sem_give(&priv->write_lock);
 
 	return (rc == kStatus_Success) ? 0 : -EINVAL;
 }
 
+#ifdef CONFIG_SW_MANAGED_FLASH_WRITE
+static int flash_mcux_write_unsafe(struct device *dev, off_t offset,
```

Comment:
When would someone ever want to define this? If power is lost, it can result in drastic corruption of flash. I would argue that the code using the flash driver always needs to know about the write alignment, and this code should just not be present.

Reply:
@d3zd3z the point of this code is to have a simple and independent code from file systems that can be used to test flash writing.
_E.g. samples/drivers/flash_shell_
```
uart:~$ flash page_read 180 20
ff ff ff ff | ff ff ff ff
ff ff ff ff | ff ff ff ff
ff ff ff ff 
uart:~$ 
uart:~$ 
uart:~$ 
uart:~$ flash page_write 180 10 1 2 3
Reading back written bytes:
01 02 03 
uart:~$ 
uart:~$ 
uart:~$ 
uart:~$ flash page_read 180 20
ff ff ff ff | ff ff ff ff
ff ff 01 02 | 03 ff ff ff
ff ff ff ff 
uart:~$ 
```

Without this code writing 512 bytes would not be possible in console, hacking code would be needed to run this sample.
Any flash programming without provisioning can lead to data corruption in case of power loss.
This function can be used as a generic low level flash write for custom sizes, I would move it to _flash.h_ .
I would just the following warnings:
 - use at your own risk
 - should not be used often, may lead to data corruption
 - when using ensure you have a stable power source

Reply:
But what real purpose does it have? I would argue that the flash shell should be updated to understand how flash devices work. We shouldn't be hiding the behavior of the underlying flash device, especially in a manner that results in corruption.

It _definitely_ should not be a general feature of the flash API.